src/Symfony/Component/Messenger/Tests/QueryBusTraitTest.php

```diff
@@ -0,0 +1,94 @@
+     * @expectedExceptionMessage You must provide a "Symfony\Component\Messenger\MessageBusInterface" instance in the "Symfony\Component\Messenger\Tests\TestQueryBusTrait::$queryBus" property, "NULL" given.
+     */
+    public function testItThrowsOnNoMessageBusInstance()
+    {
+        $trait = new TestQueryBusTrait(null);
+        $query = new DummyMessage('Hello');
+
+        $trait->dispatchQuery($query);
+    }
+
+    public function testDispatchQueryReturnsHandledStampResult()
+    {
+        $bus = $this->createMock(MessageBus::class);
+        $trait = new TestQueryBusTrait($bus);
+
+        $query = new DummyMessage('Hello');
+        $bus->expects($this->once())->method('dispatch')->willReturn(
+            new Envelope($query, new HandledStamp('result', 'DummyHandler::__invoke'))
+        );
+
+        $this->assertSame('result', $trait->dispatchQuery($query));
+    }
+
+    public function testDispatchQueryAcceptsEnvelopes()
+    {
+        $bus = $this->createMock(MessageBus::class);
+        $trait = new TestQueryBusTrait($bus);
+
+        $envelope = new Envelope(new DummyMessage('Hello'), new HandledStamp('result', 'DummyHandler::__invoke'));
+        $bus->expects($this->once())->method('dispatch')->willReturn($envelope);
+
+        $this->assertSame('result', $trait->dispatchQuery($envelope));
+    }
+
+    /**
+     * @expectedException \Symfony\Component\Messenger\Exception\LogicException
+     * @expectedExceptionMessage Query of type "Symfony\Component\Messenger\Tests\Fixtures\DummyMessage" wasn't handled. Is the "Symfony\Component\Messenger\Middleware\HandleMessageMiddleware" middleware properly registered in your bus?
+     */
+    public function testDispatchQueryThrowsOnNoHandledStamp()
+    {
+        $bus = $this->createMock(MessageBus::class);
+        $trait = new TestQueryBusTrait($bus);
+
+        $query = new DummyMessage('Hello');
+        $bus->expects($this->once())->method('dispatch')->willReturn(new Envelope($query));
+
+        $trait->dispatchQuery($query);
+    }
+
+    /**
+     * @expectedException \Symfony\Component\Messenger\Exception\LogicException
+     * @expectedExceptionMessage Query of type "Symfony\Component\Messenger\Tests\Fixtures\DummyMessage" was handled multiple times (responding handlers: "FirstDummyHandler::__invoke", "dummy_2").
+     */
+    public function testDispatchQueryThrowsOnMultipleHandledStamps()
+    {
+        $bus = $this->createMock(MessageBus::class);
+        $trait = new TestQueryBusTrait($bus);
+
+        $query = new DummyMessage('Hello');
+        $bus->expects($this->once())->method('dispatch')->willReturn(
+            new Envelope($query, new HandledStamp('first_result', 'FirstDummyHandler::__invoke'), new HandledStamp('second_result', 'SecondDummyHandler::__invoke', 'dummy_2'))
+        );
+
+        $trait->dispatchQuery($query);
+    }
+}
+
+class TestQueryBusTrait
+{
+    use QueryBusTrait {
+        dispatchQuery as public;
```

Comment:
I'd suggest the same as before: keep the method private and add a "public query(): string" ;)

Reply:
That's for tests and we just want to test `QueryBusTrait ::dispatchQuery`. Everything regarding the return type is performed by assertions.
I don't see the need to showcase faked semantics on a `QueryBusTraitConsumer` which is neither a query bus (fixed return type), neither a real trait user (the method won't be called publicly in real use-cases).

Reply:
I explained above: a QueryBus implementation won't have a fixed return type (here, `string` as you asked). Also this are tests. We use assertions on tested methods returned values.

But fine I'll changed it back.

Reply:
>  a QueryBus implementation won't have a fixed return type

I don't agree with that: it's perfectly fine and better to create typed query buses! I would recommend *never* creating untyped query buses actually.